nntrainer/graph/network_graph.cpp

```diff
@@ -0,0 +1,424 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file    network_graph.h
+ * @date    19 Oct 2020
+ * @see     https://github.com/nnstreamer/nntrainer
+ * @author  Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug     No known bugs except for NYI items
+ * @brief   This is Network Graph Class for Neural Network
+ *
+ */
+
+#include <layer_factory.h>
+#include <network_graph.h>
+
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <parse_util.h>
+
+namespace nntrainer {
+
+void NetworkGraph::addEdge(unsigned int ith, LayerNode node) {
+  if (ith > num_node - 1)
+    throw std::invalid_argument("Exceed total number of layer");
+
+  adj[ith].push_back(node);
+}
+
+void NetworkGraph::addLayerNode(std::shared_ptr<Layer> layer) {
+  std::list<LayerNode> l;
+  LayerNode *node = new LayerNode();
+
+  node->layer = layer;
+  node->index = num_node;
+
+  l.assign(1, *node);
+  adj.push_back(l);
+  num_node++;
+}
+
+void NetworkGraph::topologicalSortUtil(unsigned int ith, bool visited[],
+                                       std::stack<LayerNode> &Stack) {
+  visited[ith] = true;
+
+  std::list<LayerNode>::iterator i;
+  for (i = adj[ith].begin(); i != adj[ith].end(); ++i) {
+    if (!visited[(*i).index])
+      topologicalSortUtil((*i).index, visited, Stack);
+  }
+
+  Stack.push(getLayerNode(ith));
+}
+
+LayerNode &NetworkGraph::getLayerNode(unsigned int ith) {
+
+  std::list<LayerNode>::iterator iter;
+  for (unsigned int i = 0; i < num_node; ++i) {
+    iter = adj[i].begin();
+    if ((*iter).index == ith) {
+      return *iter;
+    }
+  }
+
+  throw std::invalid_argument("Cannot find Layer");
+}
+
+void NetworkGraph::topologicalSort() {
+  std::stack<LayerNode> Stack;
+
+  bool *visited = new bool[num_node];
+  for (unsigned int i = 0; i < num_node; ++i)
+    visited[i] = false;
+
+  for (unsigned int i = 0; i < num_node; ++i) {
+    if (visited[i] == false) {
+      topologicalSortUtil(i, visited, Stack);
```

Comment:
topologicalSortUtil() call here MUST only be made for nodes which represent input nodes. Otherwise, this implementation would result in wrong solution if input layer is not the node at 0.

Reply:
That's what I'm saying. input layer must be root and node 0. If we have to support multiple inputs then, we need update graph sort.

Reply:
Then please add this a note that not only just 1 input layer is supported but also that the input layer must be the first layer to be specified.

Also, if only 1 input layer is supported, then do we need the loop at line 75. All the nodes must be accessible starting from the input node. If not, we have a problem than starting topologicalSortUitl from there.

Reply:
That's why I left the note like
```After make node list of graph, we have to find root. (That means it should be the only one input for now.).
Need to support multiple input and support search.```

Reply:
thanks.